core/src/main/java/org/apache/iceberg/rest/HTTPRequest.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.rest;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.net.URI;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.immutables.value.Value;
+
+/** Represents an HTTP request. */
+@Value.Style(redactedMask = "****", depluralize = true)
+@Value.Immutable
+@SuppressWarnings({"ImmutablesStyle", "SafeLoggingPropagation"})
+public interface HTTPRequest {
+
+  enum HTTPMethod {
+    GET,
+    HEAD,
+    POST,
+    DELETE
+  }
+
+  /**
+   * Returns the base URI configured at the REST client level. The base URI is used to construct the
+   * full {@link #requestUri()}.
+   */
+  URI baseUri();
+
+  /**
+   * Returns the full URI of this request. The URI is constructed from the base URI, path, and query
+   * parameters. It cannot be modified directly.
+   */
+  @Value.Lazy
+  default URI requestUri() {
+    return RESTUtil.buildRequestUri(this);
+  }
+
+  /** Returns the HTTP method of this request. */
+  HTTPMethod method();
+
+  /** Returns the path of this request. */
+  String path();
+
+  /** Returns the query parameters of this request. */
+  Map<String, String> queryParameters();
+
+  /** Returns the headers of this request. */
+  @Value.Default
+  default HTTPHeaders headers() {
```

Comment:
@danielcweeks @nastra FYI, I wasn't satisfied with returning `Map<String, List<String>>` from here, especially because of corner cases around case-sensitivity. I was planning to tackle that as a follow-up task, but since we are starting from square zero here, what you see here is the "enhanced" version.